### ISSUE TITLE
Translator::getLocale() is NULL during a 404 exception handling

### ISSUE THREAD

**Issue by iltar** [symfony/symfony#13887]
I have registered a listener that fires on `kernel.exception`. This listener does some error checking and if it's a 404 error it will redirect and set a flash message. This flash message is translated before putting it in the flash message. 

The problem is that when I call `$translator->trans($key)` during the 404 exception handling, the `Translator::getLocale()` returns null. It sort of makes sense that this is not available yet due to the way the `LocaleListener` registered:
```php
    public static function getSubscribedEvents()
    {
        return array(
            // must be registered after the Router to have access to the _locale
            KernelEvents::REQUEST => array(array('onKernelRequest', 16)),
            KernelEvents::FINISH_REQUEST => array(array('onKernelFinishRequest', 0)),
        );
    }
```

This is still a bit inconvenient because I already have the locale available at this point and I don't use `_locale`. My temporary solution is to hardcode the locale in the `trans()` arguments.

_Reading this again, makes me wonder if this is the proper solution. The reason I do it in here, is because I don't want to set a message if the translation doesn't exist. My language string is build as `http.error.<status>.`_

**Comment by stof** [comment]
I suggest initializing the Translator with the default locale rather than ``null``, so that it still has a locale before the kernel.request event.

What do you think @aitboudad @fabpot ?

**Comment by aitboudad** [comment]
@stof or adding `KernelEvents::EXCEPTION` event ?
```php
    public function onKernelException(GetResponseForExceptionEvent $event)
    {
        if (null === $this->translator->getLocale()) {
            $this->setLocale($event->getRequest());
        }
    }
```

**Comment by stof** [comment]
@aitboudad this would not change the fact that we would have a translator with the ``null`` locale before actually setting it. And ``null`` is invalid (you cannot use the translator in this case)

**Comment by aitboudad** [comment]
fixed #14381
